### PR TITLE
Add ComponentHolding to packets for fully support automatic translation

### PR DESCRIPTION
**Pull request by GreenpixDev**
The "automatic component translation" option only works for packets that inherit **ComponentHoldingServerPacket** interface, so I added this interface to all packets that contain components and items. Also I added inheritance **ComponentHolder** for **ItemStack**.

### CODE REVIEW

File: src/main/java/net/minestom/server/network/packet/server/play/ResourcePackSendPacket.java [hunk 1]
```diff
@@ -56,4 +63,23 @@ public void read(@NotNull BinaryReader reader) {
...
         return ServerPacketIdentifier.RESOURCE_PACK_SEND;
     }
+
+    @Override
+    public @NotNull Collection<Component> components() {
+        List<Component> components = new ArrayList<>();
+        if (forcedMessage != null) {
+            components.add(forcedMessage);
+        }
+        return components;
```
Comment (kezz) [inline src/main/java/net/minestom/server/network/packet/server/play/ResourcePackSendPacket.java:73]: Saves an allocation I suppose!
```suggestion
        if (forcedMessage != null) {
            return Collections.singleton(forcedMessage);
        } else {
            return Collections.emptyList();
        }
```

**Review by kezz — CHANGES_REQUESTED** [review]
Could we do away with `ComponentHoldingServerPacket` entirely and just use `ComponentHolder`?

**Comment by TheMode** [comment]
A bit off-topic but I am personally still against the whole adventure translation API. They prevent a lot of optimization we are trying to push using immutable data.
Even if it gets merged, we will not be able to enforce it. (an example would be caching window items packets)

**Comment by GreenpixDev** [comment]
Indeed they prevent a lot of optimization but some users also need server flexibility. Everyone can enable or disable automatic translation or caching of packets.
If the automatic translation is not planned to be supported in the future, I suggest making it deprecated.

**Comment by TheMode** [comment]
> Indeed they prevent a lot of optimization but some users also need server flexibility. Everyone can enable or disable automatic translation or caching of packets.
> If the automatic translation is not planned to be supported in the future, I suggest making it deprecated.

The issue is that it drastically complicates the implementation of those optimizations.

We are not sure what direction to take with it, my take would be to handle all those packet transformations on a proxy. https://github.com/TheMode/Paxy

**Comment by GreenpixDev** [comment]
Ok i got it, thanks!
I close this pull request.

**Pull request closed without merging**